src/SAWScript/SolverCache.hs

```diff
@@ -0,0 +1,632 @@
+                       solverCacheMapSize = 4 {- GiB -} * 1073741824,
+                       solverCacheTimeout = 1 {- sec -} * 1000000 }
+
+-- | Create a 'SolverCache' with the given 'FilePath' and open an LMDB database
+-- at that path (i.e. `solverCacheEnv` and `solverCacheDB` are both 'Just')
+openSolverCache :: FilePath -> IO SolverCache
+openSolverCache path = do
+  (_, _, cache') <- forceSolverCacheOpened =<< lazyOpenSolverCache path
+  return cache'
+
+-- | Ensure that the given 'SolverCache' has opened an LMDB database at its set
+-- 'FilePath' - returning either the newly created or previously created
+-- 'LMDB.Environment' and 'LMDB.Database', as well as an 'SolverCache'
+-- containing both of them
+forceSolverCacheOpened :: SolverCache ->
+                          IO (LMDB.Environment LMDB.ReadWrite,
+                              LMDB.Database SolverCacheKey SolverCacheValue,
+                              SolverCache)
+forceSolverCacheOpened cache@SolverCache{ solverCacheEnv = Just env
+                                        , solverCacheDB  = Just db } =
+  return (env, db, cache)
+forceSolverCacheOpened cache@SolverCache{..} = do
+  createDirectoryIfMissing True solverCachePath
+  let limits = LMDB.defaultLimits { LMDB.mapSize = solverCacheMapSize }
+  env <- LMDB.openEnvironment solverCachePath limits
+  db <- LMDB.readOnlyTransaction env (LMDB.getDatabase Nothing)
+  let cache' = cache { solverCacheEnv = Just env, solverCacheDB = Just db }
+  return (env, db, cache')
+
+-- | Try to call 'forceSolverCacheOpened' then try to perform the given
+-- 'LMDB.Transaction' on the LMDB database associated to the given
+-- 'SolverCache', returning 'Left' if an error or timeout occurred at any point
+-- and 'Right' otherwise, as well as the updated 'SolverCache' returned by
+-- 'forceSolverCacheOpened'
+tryTransaction :: (LMDB.Mode tmode, LMDB.SubMode LMDB.ReadWrite tmode) =>
+                  SolverCache ->
+                  (LMDB.Database SolverCacheKey SolverCacheValue ->
+                   LMDB.Transaction tmode a) ->
+                  IO (Either String a, SolverCache)
+tryTransaction cache@SolverCache{..} t =
+  tryWithTimeout solverCacheTimeout (forceSolverCacheOpened cache) >>= \case
+    Right (env, db, cache') -> 
+      (,cache') <$> tryWithTimeout solverCacheTimeout
+                                   (LMDB.transaction env (t db))
+    Left err ->
+      return (Left $ "Failed to open LMDB database: " ++ err, cache)
+
+-- | An operation on a 'SolverCache', returning a value of the given type or
+-- an optional default value in the case of no enabled 'SolverCache', as well as
+-- an updated 'SolverCache'
+type SolverCacheOp a = (Maybe a, Options -> SolverCache -> IO (a, SolverCache))
+
+-- | Lookup a 'SolverCacheKey' in the solver result cache
+lookupInSolverCache :: SolverCacheKey -> SolverCacheOp (Maybe SolverCacheValue)
+lookupInSolverCache k = (Just Nothing,) $ \opts cache@SolverCache{..} ->
```

Comment:
I find this code hard to understand since there are multiple layers of `Maybe`s involved. There are three possibilities—`Nothing`, `Just Nothing`, and `Just (Just x)`—and it's unclear to me what the distinction between the three are. Why do we pick `Just Nothing` here instead of `Nothing`, for instance?

At the very least, we should document the conventions at play. It might also be worth turning `Maybe SolverCacheValue` into a bespoke data type to make it more obvious what each of the states mean.

Reply:
It think the problem was not the `Maybe SolverCacheValue`, but in fact the way I defined `SolverCacheOp`. I changed the definition of `SolverCacheOp` in bc446cf0ce29ad095e44d702ea6264e409337095 to remove the extra `Maybe` and also make things hopefully a lot more clear. This relevant bits are:
```haskell
-- | An operation on a 'SolverCache', returning a value of the given type as
-- well as an updated 'SolverCache' ('solverCacheOp'). Additionally, in the case
-- of no enabled solver cache, the operation could either fail or return a
-- default value ('solverCacheOpDefault').
data SolverCacheOp a = SCOpOrFail (Options -> SolverCache -> IO (a, SolverCache))
                     | SCOpOrDefault a (Options -> SolverCache -> IO (a, SolverCache))

-- | Get the operation associated to a 'SolverCacheOp'
solverCacheOp :: SolverCacheOp a -> Options -> SolverCache -> IO (a, SolverCache)
...

-- | Get the default value associated to a 'SolverCacheOp', if any
solverCacheOpDefault :: SolverCacheOp a -> Maybe a
...

-- | Lookup a 'SolverCacheKey' in the solver result cache
lookupInSolverCache :: SolverCacheKey -> SolverCacheOp (Maybe SolverCacheValue)
lookupInSolverCache k = SCOpOrDefault Nothing $ \opts cache@SolverCache{..} ->
...
```
Does that help make this less confusing?

Reply:
Much better! My brain started to make the Windows 95 shutdown noise when I read "`Just Nothing`", and I'm much happier now that that isn't there :)